src/Storages/Rocksdb/StorageEmbeddedRocksdb.cpp

```diff
@@ -0,0 +1,316 @@
+        auto status = storage.rocksdb_ptr->Write(rocksdb::WriteOptions(), &batch);
+        if (!status.ok())
+            throw Exception("Rocksdb write error: " + status.ToString(), ErrorCodes::SYSTEM_ERROR);
+    }
+
+private:
+    StorageEmbeddedRocksdb & storage;
+    StorageMetadataPtr metadata_snapshot;
+};
+
+StorageEmbeddedRocksdb::StorageEmbeddedRocksdb(const StorageID & table_id_,
+        const String & relative_data_path_,
+        const StorageInMemoryMetadata & metadata_,
+        bool attach,
+        Context & context_,
+        const String & primary_key_)
+    : IStorage(table_id_), primary_key{primary_key_}
+{
+    setInMemoryMetadata(metadata_);
+    rocksdb_dir = context_.getPath() + relative_data_path_ + "/rocksdb";
+    if (!attach)
+    {
+        Poco::File(rocksdb_dir).createDirectories();
+    }
+    initDb();
+}
+
+void StorageEmbeddedRocksdb::truncate(const ASTPtr &, const StorageMetadataPtr & , const Context &, TableExclusiveLockHolder &)
+{
+    rocksdb_ptr->Close();
+    Poco::File(rocksdb_dir).remove(true);
+    Poco::File(rocksdb_dir).createDirectories();
+    initDb();
+}
+
+void StorageEmbeddedRocksdb::initDb()
+{
+    rocksdb::Options options;
+    rocksdb::DB * db;
+    options.create_if_missing = true;
+    rocksdb::Status status = rocksdb::DB::Open(options, rocksdb_dir, &db);
+
+    if (status != rocksdb::Status::OK())
+        throw Exception("Fail to open rocksdb path at: " +  rocksdb_dir, ErrorCodes::SYSTEM_ERROR);
+    rocksdb_ptr = std::unique_ptr<rocksdb::DB>(db);
+}
+
+
+Pipe StorageEmbeddedRocksdb::read(
+        const Names & column_names,
+        const StorageMetadataPtr & metadata_snapshot,
+        const SelectQueryInfo & query_info,
+        const Context & /*context*/,
+        QueryProcessingStage::Enum /*processed_stage*/,
+        size_t /*max_block_size*/,
+        unsigned /*num_streams*/)
+{
+    metadata_snapshot->check(column_names, getVirtuals(), getStorageID());
+    Block sample_block = metadata_snapshot->getSampleBlock();
+    MutableColumns columns = sample_block.cloneEmptyColumns();
+
+    FieldVector keys;
+    bool all_scan = false;
+    std::tie(keys, all_scan) = getFilterKeys(primary_key, query_info.query);
+
+    // TODO pipline
+    if (all_scan)
+    {
+        auto it = std::unique_ptr<rocksdb::Iterator>(rocksdb_ptr->NewIterator(rocksdb::ReadOptions()));
+        for (it->SeekToFirst(); it->Valid(); it->Next())
+        {
+            ReadBufferFromString key_buffer(it->key());
+            ReadBufferFromString value_buffer(it->value());
+
+            for (const auto [idx, column_type] : ext::enumerate(sample_block.getColumnsWithTypeAndName()))
+            {
+                if (column_type.name == primary_key)
+                    column_type.type->deserializeBinary(*columns[idx], key_buffer);
+                else
+                    column_type.type->deserializeBinary(*columns[idx], value_buffer);
+            }
+        }
+        assert(it->status().ok());
```

Comment:
It is not a hot path, better to throw LOGICAL_ERROR to trigger the problem in release build too